diff --git a/hts.c b/hts.c
index a4dcba2..d453ad9 100644
--- a/hts.c
+++ b/hts.c
@@ -4249,6 +4249,8 @@ hts_itr_t *hts_itr_regions(const hts_idx_t *idx, hts_reglist_t *reglist, int co
                 if (itr->reg_list[i].tid < 0) {
                     if (itr->reg_list[i].tid < -1) {
                         hts_log_error("Failed to parse header");
+                        itr->reg_list = NULL;
+                        itr->n_reg = 0;
                         hts_itr_destroy(itr);
                         return NULL;
                     } else {
@@ -4262,6 +4264,8 @@ hts_itr_t *hts_itr_regions(const hts_idx_t *idx, hts_reglist_t *reglist, int co
         qsort(itr->reg_list, itr->n_reg, sizeof(hts_reglist_t), compare_regions);
         if (itr_specific(idx, itr) != 0) {
             hts_log_error("Failed to create the multi-region iterator!");
+            itr->reg_list = NULL;
+            itr->n_reg = 0;
             hts_itr_destroy(itr);
             itr = NULL;
         }
diff --git a/htslib/hts.h b/htslib/hts.h
index f9c0a9a..f5636ca 100644
--- a/htslib/hts.h
+++ b/htslib/hts.h
@@ -1347,8 +1347,9 @@ typedef int hts_itr_multi_query_func(const hts_idx_t *idx, void *itr);
     @param tell         Callback to return current input file location
     @return An iterator on success; NULL on failure

-    The iterator struct returned by a successful call should be freed
-    via hts_itr_destroy() when it is no longer needed.
+    On success the iterator owns @p reglist and should be freed via
+    hts_itr_destroy() when it is no longer needed.  On failure ownership of
+    @p reglist remains with the caller.
  */
 HTSLIB_EXPORT
 hts_itr_t *hts_itr_regions(const hts_idx_t *idx, hts_reglist_t *reglist, int count, hts_name2id_f getid, void *hdr, hts_itr_multi_query_func *itr_specific, hts_readrec_func *readrec, hts_seek_func *seek, hts_tell_func *tell);
diff --git a/htslib/sam.h b/htslib/sam.h
index 6d85b30..b74c780 100644
--- a/htslib/sam.h
+++ b/htslib/sam.h
@@ -1333,6 +1333,9 @@ in.

 The iterator will return all reads overlapping the given regions.  If a read
 overlaps more than one region, it will only be returned once.
+
+On success the iterator owns @p reglist and should be freed with
+hts_itr_destroy().  On failure ownership of @p reglist remains with the caller.
  */
 HTSLIB_EXPORT
 hts_itr_t *sam_itr_regions(const hts_idx_t *idx, sam_hdr_t *hdr, hts_reglist_t *reglist, unsigned int regcount);
